GITHUB#11795: Add FilterDirectory to track write amplification factor

mdmarshmallow opened the pull request:
Added a `WriteAmplificationTrackingDirectoryWrapper` that simply keeps track of bytes flushed and bytes merged. This allows us to get an estimate of the write amplification factor by doing `(bytesFlushed + bytesMerged) / bytesFlushed)`.

Commit:
LUCENE-11795: Add FilterDirectory to track write amplification factor

gautamworah96 reviewed (commented):
I tried to make a first pass at this. I have some comments on alternative ideas for the general approach as well, and have not looked the tests. I think we are moving in the correct direction.

gautamworah96 commented on lucene/core/src/java/org/apache/lucene/store/WriteAmplificationTrackingDirectoryWrapper.java:
```diff
@@ -0,0 +1,59 @@
+  public IndexOutput createOutput(String name, IOContext context) throws IOException {
+    IndexOutput output = in.createOutput(name, context);
+    IndexOutput byteTrackingIndexOutput;
+    if (context.context.equals(IOContext.Context.FLUSH)) {
```
`context.context` is a bit confusing. Lets rename the method param?

mdmarshmallow replied:
Will do.

mdmarshmallow commented on lucene/core/src/java/org/apache/lucene/store/ByteTrackingIndexOutput.java:
```diff
@@ -0,0 +1,70 @@
+    return output.getChecksum();
+  }
+
+  public String getWrappedName() {
```
`super#getName()` and `super#toString()` would give us the name and String representation of this class. I made these in case someone wants the access the wrapped class's name and String representation. With that being said, I'm not really sure how useful these two methods would be. I could remove them?

mdmarshmallow commented on lucene/core/src/java/org/apache/lucene/store/WriteAmplificationTrackingDirectoryWrapper.java:
```diff
@@ -0,0 +1,59 @@
+    return byteTrackingIndexOutput;
+  }
+
+  public double getApproximateWriteAmplificationFactor() {
```
That's true, but is there ever a case where we would have two directories open for some reason or another? Then we'd have to think about this a bit more if we want to merge the write amplification factors or not.

vigyasharma reviewed (commented):
Thanks for making these changes, @mdmarshmallow! 

We've been simulating write amplification by writing known file sizes in tests. This should make it easier to measure it with actual input data (like in luceneutil benchmarks).

jpountz commented on lucene/core/src/java/org/apache/lucene/store/ByteTrackingIndexOutput.java:
```diff
@@ -0,0 +1,70 @@
+public class ByteTrackingIndexOutput extends IndexOutput {
+
+  private final IndexOutput output;
+  private final AtomicLong byteTracker;
```
IndexOutput is indeed not thread-safe. I think that the difference between this class and the other ones you referred to is that this one shares its counter across multiple output instances, so the counter needs to be thread-safe.

vigyasharma replied:
Ah I see what you mean.. The counters here are passed to it in the ctor, and may be shared across multiple IndexOutputs.. Makes sense..

vigyasharma replied:
I feel there's a recurring need to time/measure/count metrics across different parts of Lucene. It might be a good idea to add some Stats object and interface to Lucene. I'll open an issue to discuss and frame this idea.

jpountz commented:
This class feels like it'd be a good fit for the `misc` module rather than `core`?

mikemccand commented:
I love this approach/idea!

It's simple so we should start with this ... but it will necessarily be a lagging indicator since merging takes some time to kick off and run to completely while flushing keeps happening if docs are being indexed.  Also, it reports the "for all time" WAF instead of adding some decay and being closer to an instantaneous measure.  But we can try to improve those later.

Have you tried turning this on for a `luceneutil` indexing run to see what WAF it reports?  It might be tricky to do it right because by default I think `luceneutil` indexing does not wait for merges while indexing.

mikemccand commented:
An alternative implementation would be to add the bytes only in the `IndexOutput.close` method instead of on each method that writes bytes?  It might be less error-proned, but, also less real-time since it won't be until the file is closed that we count any bytes in the shared counters.

mikemccand commented on lucene/core/src/java/org/apache/lucene/store/ByteTrackingIndexOutput.java:
```diff
@@ -0,0 +1,70 @@
+  }
+
+  @Override
+  public void close() throws IOException {
```
Maybe we can separately add a proper / tested delegator, `FilterIndexOutput`.  Lucene has a number of these delegator classes (`FilterXXX`) but not yet `FilterIndexOutput`.  Elasticsearch seems to have one... but if you poach/borrow from there make sure to use the 7.10 or earlier version that is still Apache Software License.

mdmarshmallow replied:
That makes sense to me. I'll make a separate issue to track since at a quick glance, there are other `XXXIndexOutput` classes that might need to be changed.

mdmarshmallow commented on lucene/core/src/java/org/apache/lucene/store/ByteTrackingIndexOutput.java:
```diff
@@ -0,0 +1,70 @@
+
+  @Override
+  public void writeBytes(byte[] b, int offset, int length) throws IOException {
+    byteTracker.addAndGet(length);
```
Yeah this makes sense, will change the order here.,

mdmarshmallow commented on lucene/core/src/java/org/apache/lucene/store/ByteTrackingIndexOutput.java:
```diff
@@ -0,0 +1,70 @@
+  public String getWrappedToString() {
+    return output.toString();
+  }
+}
```
Hmm, I think if those methods get overridden though, that would break this implementation cause it would use the wrapped `IndexOutput#writeBytes` in which case we won't be tracking the bytes anymore I think?

vigyasharma replied:
Isn't the reverse true.. overriding those functions will help you continue to track those bytes? 
e.g. If I wrap `OutputStreamIndexOutput` with `ByteTrackingIndexOutput` today, and call `writeShort()` or `writeInt()`, won't I lose tracking information?

mdmarshmallow replied:
Oh I see what you're saying, I got a bit confused here. I'll change this.

mikemccand replied:
I think this is why we need `FilterIndexOutput`?  To make delegation always 100%.

mdmarshmallow commented:
So by doing this on `IndexOutput.close()`, we would avoid including half-done merges/flushes in the write amplification factor? As you said, this does track all-time WAF so I guess being real-time is not as much of a concern.

vigyasharma commented:
I see you'd already responded to a bunch of my comments. I should've refreshed my PR page. Will resolve those.

gautamworah96 commented:
For folks more familiar with WAF calculations for Search applications, is the formula of `(flushedBytes + mergedBytes) / flushedBytes` always correct? 

For example, does the `IOContext.Context.MERGE` operation not include all the bytes written during a `FLUSH` operation (i.e when we are writing to disk)? or should it be something like `mergedBytes/flushedBytes` when there have been merges and `1` otherwise when `flushedBytes` are 0?

Commit:
addressed feedback

vigyasharma commented:
> An alternative implementation would be to add the bytes only in the `IndexOutput.close` method instead of on each method that writes bytes? It might be less error-proned, but, also less real-time since it won't be until the file is closed that we count any bytes in the shared counters.

I'm a bit conflicted about this. I like the completeness we get after `close()` is called. But as an API, consumers now have to be careful with `getApproximateWriteAmplificationFactor()`.. 

There is nothing stopping them from calling it before IndexOutput is closed. The counter will just return 0. Or the value it held before it was reused. The onus of ensuring close() was called is on the caller here i.e. the dir wrapper.

However, in the dir wrapper, we don't keep references to different IndexOutputs for flush and merge. We directly read the counter values in getApproximateWriteAmplificationFactor(), and so there's no way to throw an error, if someone calls write amplification early. 

In other words, we cannot ensure that when write amplification returns 1, it really is 1. It could be because IndexOutput is still open.

Maybe, we should let BytesTrackingIndexOutput expose a `bytesWritten()` method, which internally verifies that close was called. A subsequent real-time writes impl. could change this. The dir. wrapper would then keep IndexOutput references around, and use them instead of directly reading counters. 

Then we don't need to pass shared Atomic counters. We can directly aggregate values across IndexOutput references if we want to.

vigyasharma commented:
Thanks for persisting with this @mdmarshmallow. I think we're close now, just a couple of discussion threads to resolve. This change will be super useful :)

jpountz commented:
This implementation ignores temporary index outputs from write amplification, which I wonder whether this is correct (maybe it is, I struggle making an opinion on this question).

vigyasharma commented:
> This implementation ignores temporary index outputs from write amplification, which I wonder whether this is correct (maybe it is, I struggle making an opinion on this question).

Interesting point.. Thinking how/when we'd like to track the impact of temp output files. From what I understand, they won't be a part of commit and fsync. So if we're trying to measure increased disk or remote store I/O, we probably want to skip them?

Maybe, when we make optimizations that write more temp files (like #11411 ?), we'll use this to measure some impact. Although we delete the temp files right after, but on a small box, maybe we gives us a sense of increased file writes or page fault.

We could add a flag to optionally include temp files.. It would require overriding `createTempOutput()` right?

dsmiley commented:
This is nifty!  
I wonder if it'd be worthwhile for Lucene itself to track this small bit of metadata so that it's persistent?

Commit:
added optional temp output tracking and real time tracking

mdmarshmallow commented:
Sorry for the delayed response, I haven't had internet for a few days. I added the option to include temp output in tracking, as well as the a method for real time byte tracking as well so we can let users get exactly what they need.

Also @dsmiley that's an interesting suggestion! I'm not as familiar with Lucene as some of the other people commenting here but I would be open to adding this to metadata if other's agreed it would be useful.

jpountz commented:
> Interesting point.. Thinking how/when we'd like to track the impact of temp output files. From what I understand, they won't be a part of commit and fsync. So if we're trying to measure increased disk or remote store I/O, we probably want to skip them?

Indeed temporary files are never part of a commit and fsynced, but this may also be the case for a number of flushed segments: if flushed segments get included in a merge before the next commit, then they would never be part of a commit and fsynced either.

> Although we delete the temp files right after, but on a small box, maybe we gives us a sense of increased file writes or page fault.

Some temporary files are also not necessarily short-lived, like the ones we create for stored fields when index sorting is enabled.

I'm considering exposing write amplification separately for flushes (as `flushedBytes / totalIndexSize`), merges (as `(totalIndexSize + mergedBytes) / totalIndexSize`) and temporary files (as `(totalIndexSize + tempBytes) / totalIndexSize`) and pushing the responsibility to users of whether and how they would like to combine these various metrics?

zhaih commented on lucene/misc/src/java/org/apache/lucene/misc/store/ByteWritesTrackingDirectoryWrapper.java:
```diff
@@ -0,0 +1,118 @@
+    return (flushedBytes + mergedBytes) / flushedBytes;
+  }
+
+  /** Gets a more up-to-date but less accurate write amplification factor */
```
I wonder why this is less accurate? Seems to me it is more frequently updated and thus should be more accurate?

mdmarshmallow replied:
My understanding is that the real time metric would track partially completed flushed and merges, so it wouldn't be the "true" write amplification factor of all completed flushes and merges. I'm not sure if there are any cases where a flush or merge would get cancelled before it's completed though to be honest.

zhaih replied:
OK, in case any file are cancelled then this one will be inaccurate, makes sense

zhaih replied:
Maybe add the explanation to javadoc?

zhaih commented on lucene/misc/src/java/org/apache/lucene/misc/store/ByteWritesTrackingDirectoryWrapper.java:
```diff
@@ -0,0 +1,118 @@
+
+  private final AtomicLong flushedBytes = new AtomicLong();
+  private final AtomicLong mergedBytes = new AtomicLong();
+  private final AtomicLong realTimeFlushedBytes = new AtomicLong();
```
This real time counter could potentially harm the performance since it is increased every byte you write for every thread.

Since the IndexOutput is already not thread-safe, we probably can let `ByteTrackingIndexOutput` keep a private (volatile?) counter of bytes, and expose a public getter. Then in this directory wrapper keep a list of tracking output so that when we need we can just call the getter methods and sum up the numbers. Also to avoid tracking too much index output we can deregister them when they're closed.

It's just another idea and could be a follow-up, but I think we need to be careful about using the atomics.

mdmarshmallow replied:
I think that's a good idea. Not sure we need to the counter to be volatile though since each `ByteTrackingIndexOutput` operates on a single thread right? So if the counter is private it should be fine I think.

zhaih replied:
I think volatile is necessary because it need to be read by another thread possibly

mdmarshmallow replied:
Oh I think you're right. I'll make that change

uschindler replied:
Volatile is worse than AtomicLong, because it has a barrier on both read and write. So the atomic update is better.

uschindler replied:
Do we really read from another thread while writing? Updating the counter is fine as it only happens in one thread. The problem is reading so we get a value that's up-to-date. The question is how up-to-date it needs to be.

Maybe we can have a standard long value that gets updated by the writing thread. Every n~1000 bytes written we update the atomic one. What do you think?

mdmarshmallow replied:
I don't see any issue with having a real time counter lagging by 1000 bytes, though maybe we can also provide that as an option that can be changed and have the default at 1000?

uschindler replied:
Alternatively use a varhandle and mostly call get/set to update it in writer thread and setVolatile from time to time to persist value so it is visible by other threads. Other threads always use getVolatile to check.

zhaih replied:
> Volatile is worse than AtomicLong, because it has a barrier on both read and write. So the atomic update is better.

Thanks for pointing that out, new knowledge got :)
> Maybe we can have a standard long value that gets updated by the writing thread. Every n~1000 bytes written we update the atomic one. What do you think?

I'll vote for this approach, it sounds neat and I don't think a 1KB error will make too much difference, and we can make it configurable even.

uschindler replied:
I would add a local plain `int` field (an int is enough here, as we only write small amounts and it fits better to the parameters passed to the methods) to the IndexOutput, increment it on every write completely unsychronized. After every write call a private method  like `maybeUpdateGlobal()` which checks in the plain `int` field if it is larger than a certain value and then does something like `AtomicLong#addAndGet(localIntValue); localIntValue = 0;`

Nevertheless we should have a benchmark how much this affects writing. If the filter directory is just for debugging purposes, we can of course go with the full atomic implementation as we have now. Nobody takes care if you want to just watch the amplification factors while debugging your app and it gets slower because of this.

mikemccand replied:
If we switch to tracking only on `close` (and explain the small loss of real-time-ness in the javadocs), then 1) we can privately track in a local `long` the byte count written, 2) the risk to performance should be lower, 3) PR gets simpler/smaller?  We can "transfer" the `long bytesWritten` to the shared `AtomicLong` on close.

mikemccand replied:
Actually, on `close` couldn't we just call `getFilePointer()` to get the number of bytes written?

uschindler replied:
yes, Mike or Robert are right. you can just capture getFilePointer() on close(). This makes the risk of performance issues zero.
IMHO, we don't need the live information while the stuff is written. If we record it at end it should be fine.

uschindler replied:
If you are wondering why getFilePointer() works: The name getFilePointer() is a bit misnomed from former times. Nowadays you can't seek while writing to IndexOutput, the semantics are now identical to an OutputStream (it goes forward only, only one thread can write). So after you have written everything, the file pointer is just pointing after the last byte, which is the number of bytes written.

mdmarshmallow commented:
@jpountz , in response to this:

> I'm considering exposing write amplification separately for flushes (as flushedBytes / totalIndexSize), merges (as (totalIndexSize + mergedBytes) / totalIndexSize) and temporary files (as (totalIndexSize + tempBytes) / totalIndexSize) and pushing the responsibility to users of whether and how they would like to combine these various metrics?

We could add these in addition to the write amplification factor methods that are already present?

mikemccand commented:
> Also @dsmiley that's an interesting suggestion! I'm not as familiar with Lucene as some of the other people commenting here but I would be open to adding this to metadata if there are no objections.

It would indeed be awesome if this was a first class metric in Lucene.  But let's first start with this `Directory` wrapper approach and see how it's used / improved, and later upgrade it to first class if it is helpful?  Progress not perfection!

mikemccand commented:
> I'm considering exposing write amplification separately for flushes (as `flushedBytes / totalIndexSize`), merges (as `(totalIndexSize + mergedBytes) / totalIndexSize`) and temporary files (as `(totalIndexSize + tempBytes) / totalIndexSize`) and pushing the responsibility to users of whether and how they would like to combine these various metrics?

+1

uschindler reviewed (changes requested):
see my comments above

rmuir commented:
why does this need to do anything more than `getFilePointer()` on `close()` or something to capture how much was written?

uschindler commented:
Robert is right, why do we need to see the values live? getFilePointer() always works.

jpountz commented:
>> I'm considering exposing write amplification separately for flushes (as flushedBytes / totalIndexSize), merges (as (totalIndexSize + mergedBytes) / totalIndexSize) and temporary files (as (totalIndexSize + tempBytes) / totalIndexSize) and pushing the responsibility to users of whether and how they would like to combine these various metrics?
>
> We could add these in addition to the write amplification factor methods that are already present?

I think that one question I was raising was also whether the current way we're measuring write amplification is correct. Say you write 10 1MB segments, which then get merged into a 9MB segment (smaller the the sum of the sizes of flushed segments, possibly because many terms are no longer duplicated across segments). What is the write amplification of merging? Is it `(10 + 9) / 10 = 1.9` like the formula that this PR is currently using, computing write ampfilication as a function of the total flush size. Or is it `(9 + 9) / 9 = 2`, computing write amplification as a function of the total index size? The latter feels more intuitive to me.

dsmiley commented:
The former is more intuitive to me -- how much more data do we write beyond the initial segment flush.  This is the added cost of a system immutable files with log structured merge.

Commit:
addressed more feedback

mdmarshmallow commented:
I agree, the former is also more intuitive to me as well for the same reasons. In your example, 10MB were initially written, and 9MB were written again in merges, so 1.9 to me means that we did 1.9 times the work of initially just writing the 10MB to get to the final 9MB index state.

Though since this does seem like it might be a point of confusion for users in that there are multiple ways to calculate this, do you think it would be better to just expose `getFlushedBytes()` and `getMergedBytes()` and let users calculate the write amplification as they see fit?

danmuzi commented on lucene/misc/src/java/org/apache/lucene/misc/store/ByteWritesTrackingDirectoryWrapper.java:
```diff
@@ -0,0 +1,102 @@
+      throws IOException {
+    IndexOutput output = in.createTempOutput(prefix, suffix, ioContext);
+    if (trackTempOutput) {
+      ByteTrackingIndexOutput byteTrackingIndexOutput;
+      if (ioContext.context.equals(IOContext.Context.FLUSH)) {
+        byteTrackingIndexOutput = new ByteTrackingIndexOutput(output, flushedBytes);
+      } else if (ioContext.context.equals(IOContext.Context.MERGE)) {
+        byteTrackingIndexOutput = new ByteTrackingIndexOutput(output, mergedBytes);
+      } else {
+        return output;
+      }
+      return byteTrackingIndexOutput;
```
There are duplicate codes.
How about changing it to:
```java
  @Override
  public IndexOutput createOutput(String name, IOContext ioContext) throws IOException {
    IndexOutput output = in.createOutput(name, ioContext);
    return createByteTrackingOutput(output, ioContext.context);
  }

  @Override
  public IndexOutput createTempOutput(String prefix, String suffix, IOContext ioContext)
      throws IOException {
    IndexOutput output = in.createTempOutput(prefix, suffix, ioContext);
    return trackTempOutput ? createByteTrackingOutput(output, ioContext.context) : output;
  }

  private IndexOutput createByteTrackingOutput(IndexOutput output, IOContext.Context context) {
   // If you want to add this feature to Lucene 9, you should modify it to the old switch-case style.
    return switch (context) {
      case FLUSH -> new ByteTrackingIndexOutput(output, flushedBytes);
      case MERGE -> new ByteTrackingIndexOutput(output, mergedBytes);
      default -> output;
    };
  }
```

mdmarshmallow replied:
Hi, thanks for the feedback! I made the changes you suggested + added a new CHANGED.txt entry under 9.5

danmuzi reviewed (changes requested):
Thanks for making this PR, @mdmarshmallow 👍
I left a simple comment.
And you need to add a description to lucene/CHANGES.txt.

Commit:
more improvements + added CHANGED.txt entry

danmuzi commented on lucene/misc/src/java/org/apache/lucene/misc/store/ByteWritesTrackingDirectoryWrapper.java:
```diff
@@ -0,0 +1,96 @@
+        return new ByteTrackingIndexOutput(output, flushedBytes);
+      case MERGE:
+        return new ByteTrackingIndexOutput(output, mergedBytes);
+      case DEFAULT:
+      case READ:
+      default:
```
DEFAULT and READ can be handled by default.

mdmarshmallow replied:
Checkstyle complains when I don't explicitly include all cases in the enum unfortunately.

danmuzi commented on lucene/CHANGES.txt:
```diff
@@ -98,6 +98,11 @@ API Changes
+New Features
+---------------------
+* GITHUB#11795: Add ByteWritesTrackingDirectoryWrapper to expose metrics for bytes merged, flushed, and overall
+write amplification factor. (Marc D'Mello)
```
Usually, when a new line is needed for a description in CHANGES.txt, two whitespaces are added in front.
(see other multi-line description)

write amp... -> {space}{space}write amp...

danmuzi reviewed (changes requested):
Thanks for checking.
I left a few more minor comments.

mdmarshmallow commented:
Not sure why the tests are failing as `./gradlew check` passes on my machine. I checked the macos-latest distributions test logs and found this:
```
2022-09-29T19:53:49.4803130Z Failed to resolve action download info. Error: nodename nor servname provided, or not known (pipelines.actions.githubusercontent.com:443)
2022-09-29T19:53:49.4820400Z Retrying in 13.157 seconds
2022-09-29T19:55:42.7778120Z Failed to resolve action download info. Error: The HTTP request timed out after 00:01:40.
2022-09-29T19:55:42.7787910Z Retrying in 15.793 seconds
2022-09-29T19:56:49.5196790Z ##[error]Failed to resolve action download info.
```
It seems like a transient error to me, but I'm not knowledgable enough about this to be sure.

Commit:
format edit to CHANGES.txt

danmuzi reviewed (approved):
Looks good to me 👍

jpountz commented:
> Though since this does seem like it might be a point of confusion for users in that there are multiple ways to calculate this, do you think it would be better to just expose getFlushedBytes() and getMergedBytes() and let users calculate the write amplification as they see fit?

This sounds like a good idea to me.

Commit:
remove waf factor calculation

mdmarshmallow commented:
Sounds good, removed the `getApproximateWriteAmplficationFactor()` function.

mdmarshmallow committed:
Merge branch 'main' into waf-metric

mdmarshmallow commented:
I resolved a merge conflict with `CHANGES.txt`. I think I've also addressed all the issues people have pointed out as well. Would it be possible for someone to take a look at this and merge it if there are no more objections? Thanks!

mikemccand commented on lucene/misc/src/java/org/apache/lucene/misc/store/ByteTrackingIndexOutput.java:
```diff
@@ -0,0 +1,91 @@
+import org.apache.lucene.store.IndexOutput;
+
+/** An {@link IndexOutput} that wraps another instance and tracks the number of bytes written */
+public class ByteTrackingIndexOutput extends IndexOutput {
```
Maybe open a follow-on issue to add a `FilterIndexOutput`?  These delegators are spooky when they are not properly tested...

mdmarshmallow replied:
I made an issue to track this here: https://github.com/apache/lucene/issues/11868

mikemccand commented:
This looks great to me!  I love all the engagement (83+ comments!) and how it iterated to such a simple solution.  I left a small comment for a follow-on issue ... and it looks like `CHANGES.txt` is conflicting again

@mdmarshmallow maybe open another follow-on issue in `luceneutil` to add this to nightly benchmarks?  It'd be great to see impact on WAF over time of interesting index-time changes...

I'll push this in a few days if nobody objects.

mdmarshmallow committed:
Merge branch 'main' into waf-metric

mdmarshmallow commented:
Thanks Mike, I added an issue to `luceneutil`: https://github.com/mikemccand/luceneutil/issues/208

mikemccand commented:
Thanks @mdmarshmallow!  Sorry for the delay merging ... I will backport to 9.x then let's get this in nightly benchmarks :)

mikemccand commented:
9.x backport done: https://github.com/apache/lucene/commit/373d2e84c13ee67e8e1247338e69b53946b7f726
